File: Easy/python/sorts.py

def swap(i, j, arr):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp


# Bubble
arr = [8, 5, 2, 9, 5, 6, 3]


def bubbleSort(arr):
    swapped = False
    for i in range(len(arr) - 1):
        if arr[i] > arr[i + 1]:
            swap(i, i + 1, arr)
            swapped = True
    return bubbleSort(arr) if swapped else arr


print('bubble', bubbleSort(arr))

# Insertion
arr = [8, 5, 2, 9, 5, 6, 3]


def insertionSort(arr):
    for i in range(1, len(arr)):
        j = i
        while (j > 0 and arr[j] < arr[j - 1]):
            swap(j, j - 1, arr)
            j -= 1
    return arr


print('insertion', insertionSort(arr))

# Selection
arr = [8, 5, 2, 9, 5, 6, 3]


def selectionSort(arr):
    for i in range(len(arr)):
        minIndex = i
        for j in range(i + 1, len(arr)):
            if (arr[j] < arr[minIndex]):
                minIndex = j
        swap(i, minIndex, arr)
    return arr


print('selection', selectionSort(arr))

File: Easy/js/sorts.js
const array = [8, 5, 2, 9, 5, 6, 3];

// helper
const swap = (i, j, arr) => {
  let temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
};

// bubble sort
let arr = array.slice();
const bubbleSort = arr => {
  let swapped = false;
  for (let i = 0; i < arr.length - 1; i++) {
    if (arr[i] > arr[i + 1]) {
      swap(i, i + 1, arr);
      swapped = true;
    }
  }
  return swapped ? bubbleSort(arr) : arr;
};
console.log("bubble", bubbleSort(arr));

// insertion sort
arr = array.slice();
const insertionSort = arr => {
  for (let i = 1; i < arr.length; i++) {
    let j = i;
    while (j > 0 && arr[j] < arr[j - 1]) {
      swap(j, j - 1, arr);
      j--;
    }
  }
  return arr;
};

arr = array.slice();
const insertionSort2 = arr => {
  for (let i = 1; i < arr.length; i++) {
    let val = arr[i];
    for (var j = i - 1; j >= 0 && arr[j] > val; j--) {
      arr[j + 1] = arr[j];
    }
    arr[j + 1] = val;
  }
  return arr;
};

console.log("insertion", insertionSort(arr));
console.log("insertion2", insertionSort2(arr));

// selection sort
arr = array.slice();
const selectionSort = arr => {
  for (let i = 0; i < arr.length - 1; i++) {
    let minIdx = i;
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[j] < arr[minIdx]) {
        minIdx = j;
      }
    }
    swap(i, minIdx, arr);
  }
  return arr;
};

console.log("selectionSort", selectionSort(arr));
